### DISCUSSION TITLE
利用SNI分流，stream模块443流量转发后，可以富强和建立其他web网站，但是速度有很明显的下降的问题探究。

### DISCUSSION THREAD

**Discussion by ghost** [XTLS/Xray-core#697] [Q&A]
我目前使用的是VLESS+TCP+xtls-rprx-direct，网上找到的一键脚本和按步骤搭建，xray都需要占有443端口，
于是找到了利用SNI分流，stream模块加持下，解决了443端口被xray独占的情况,
但是网速大打折扣，比如若直接让xray监听443 速度快，可以跑满我本地的300Mbps电信带宽，但是不能再部署其他的网站，vps资源被浪费了有点不值得；若用了nginx stream模块，监听其他端口，443端口转发后，虽然也可以搭建其他web服务，可是安装了xray科学富强后，速度测速只有80Mmps，请问：这是nginxSNI这项技术的弊端吗，还是说我猜测可以实现相同的速度，只是多了这一步监听转发对vps性能过高，但我检查了CPU，两种情况都没有超过50%的负荷，所以特别想知道各位使用了分流共用443端口后，速度是否有明显的减弱。

**Comment by lw4free** [comment]
我也使用nginx stream 测速有超过200M 使用tcp+tls(nginx)+vless 可以回落 不能使用xtls 你是怎样配置的?
另请教 你使用nginx stream SNI的配置

**Comment by ghost** [comment]
> 我也使用nginx stream 测速有超过200M 使用tcp+tls(nginx)+vless 可以回落 不能使用xtls 你是怎样配置的?
> 另请教 你使用nginx stream SNI的配置

 这是我自己写的笔记，里面有配置文件，可以实现xtls 也是参考了网上好多别人的配置， 但最终还是没有办法实现跑慢本地带宽，怀疑是两个原因，第一，VPS的CPU单核可能分流达不到性能要求，第二采用这样子分流的方式速度有较大损耗，或许可以通过调参来改善，但是毕竟不是专业的，也搞不太懂。
[传送门](https://notes.zhangjie.work/index.php/archives/138/)

**Comment by xqdoo00o** [comment]
reuseport 好像是会变慢，去掉 reuseport 试试

**Comment by timi-owo** [comment]


**Reply by gmvp3** [reply]
不明觉厉

**Reply by Moius** [reply]
请问这个损耗大么？

**Comment by lw4free** [comment]
> 
> 
> > 我也使用nginx stream 测速有超过200M 使用tcp+tls(nginx)+vless 可以回落 不能使用xtls 你是怎样配置的?
> > 另请教 你使用nginx stream SNI的配置
> 
> 这是我自己写的笔记，里面有配置文件，可以实现xtls 也是参考了网上好多别人的配置， 但最终还是没有办法实现跑慢本地带宽，怀疑是两个原因，第一，VPS的CPU单核可能分流达不到性能要求，第二采用这样子分流的方式速度有较大损耗，或许可以通过调参来改善，但是毕竟不是专业的，也搞不太懂。
> [传送门](https://notes.zhangjie.work/index.php/archives/138/)

参考你的SNI配置成功分流，没有发现掉速，能跑300M+。
主要配置：
```
ningx:
stream {
    map $ssl_preread_server_name $name {
        mydomain                xtls;
        www.mydomain            http;
    }
    # upstream pool
    upstream xtls {
        server localhost:8081;
    }
    upstream http {
        server localhost:8080;
    }
    server {
        listen 443 so_keepalive=on;
        proxy_pass $name;
        proxy_protocol on;
        ssl_preread on;
    }
}
xray:
    {
      "tag": "xtls",
      "port": 8081,
      "listen": "127.0.0.1",
      "protocol": "vless",
      "settings": {
        "clients": [
        ],
        "decryption": "none",
        "fallbacks": [
          {
            "dest": 8080
          }
        ]
      },
      "streamSettings": {
        "network": "tcp",
        "security": "xtls",
        "xtlsSettings": {
          "alpn": [
            "http/1.1"
          ],
          "certificates": [
            {
              "certificateFile": "/etc/ssl/cert.pem",
              "keyFile": "/etc/ssl/key.pem"
            }
          ]
        },
        "tcpSettings": {
          "acceptProxyProtocol": true
        }
      }
    },
```

**Comment by ghost** [comment]
> > > 我也使用nginx stream 测速有超过200M 使用tcp+tls(nginx)+vless 可以回落 不能使用xtls 你是怎样配置的?
> > > 另请教 你使用nginx stream SNI的配置
> > 
> > 
> > 这是我自己写的笔记，里面有配置文件，可以实现xtls 也是参考了网上好多别人的配置， 但最终还是没有办法实现跑慢本地带宽，怀疑是两个原因，第一，VPS的CPU单核可能分流达不到性能要求，第二采用这样子分流的方式速度有较大损耗，或许可以通过调参来改善，但是毕竟不是专业的，也搞不太懂。
> > [传送门](https://notes.zhangjie.work/index.php/archives/138/)
> 
> 参考你的SNI配置成功分流，没有发现掉速，能跑300M+。
> 主要配置：
> 
> ```
> ningx:
> stream {
>     map $ssl_preread_server_name $name {
>         mydomain                xtls;
>         www.mydomain            http;
>     }
>     # upstream pool
>     upstream xtls {
>         server localhost:8081;
>     }
>     upstream http {
>         server localhost:8080;
>     }
>     server {
>         listen 443 so_keepalive=on;
>         proxy_pass $name;
>         proxy_protocol on;
>         ssl_preread on;
>     }
> }
> xray:
>     {
>       "tag": "xtls",
>       "port": 8081,
>       "listen": "127.0.0.1",
>       "protocol": "vless",
>       "settings": {
>         "clients": [
>         ],
>         "decryption": "none",
>         "fallbacks": [
>           {
>             "dest": 8080
>           }
>         ]
>       },
>       "streamSettings": {
>         "network": "tcp",
>         "security": "xtls",
>         "xtlsSettings": {
>           "alpn": [
>             "http/1.1"
>           ],
>           "certificates": [
>             {
>               "certificateFile": "/etc/ssl/cert.pem",
>               "keyFile": "/etc/ssl/key.pem"
>             }
>           ]
>         },
>         "tcpSettings": {
>           "acceptProxyProtocol": true
>         }
>       }
>     },
> ```

`listen 443 so_keepalive=on;`以及` "acceptProxyProtocol": true`，是我配置中所没有的，这或许就是我的不能跑满本地带宽的症结所在吧。前面熊猫大哥也提及了`reuseport 好像是会变慢`，看来真的是这个原因了。

下午折腾了一下，还是不能跑满。方便的话共享一下你的可以跑满本地带宽的nginx配置、xray配置以及伪装网站的配置文件。完整的配置信息我自己再对照一下。

**Comment by lw4free** [comment]
nginx和xray主要配置就上面那些了 其它配置应该对速度没有影响 我也没有跑满公司500M vps和透明网关比较弱鸡

**Comment by UptonEdward** [comment]
请问你的问题解决了吗？我也遇到了同样的问题，现在一筹莫展呀

**Comment by cvpas** [comment]
我也遇到这个问题，谷歌了一圈终于找到个不影响速度的nginx前置的配置，你可以试试。https://cnix.win/87.html
另外麻烦问一下，你xray配置里开启了
```
        "tcpSettings": {
          "acceptProxyProtocol": true
        }
```
请问xray error日志里是否有[Warning] transport/internet/tcp: accepting PROXY protocol的提示信息呢？

**Reply by Moius** [reply]
你好，出现这种warning有可能是你的nginx配置里没开启proxyprotocol？如果开启了且版本相同，他应该不会给报这个的，请注意，nginx只支持V1版本标识。proxyprotocol是一个用于标识用户元数据的数据头，开启他才可以在nginx分流情况下得到真实IP

**Reply by cvpas** [reply]
你好 麻烦看看这个，我在这个帖子反馈了一下。

https://github.com/XTLS/Xray-core/discussions/1012

**Comment by ghost** [comment]
我也Google了一圈，so_keepalive=on 以及不使用 reuseport 更容易导致惊群问题吧？

**Comment by taxtigerno1** [comment]
> 我目前使用的是VLESS+TCP+xtls-rprx-direct，网上找到的一键脚本和按步骤搭建，xray都需要占有443端口， 于是找到了利用SNI分流，stream模块加持下，解决了443端口被xray独占的情况, 但是网速大打折扣，比如若直接让xray监听443 速度快，可以跑满我本地的300Mbps电信带宽，但是不能再部署其他的网站，vps资源被浪费了有点不值得；若用了nginx stream模块，监听其他端口，443端口转发后，虽然也可以搭建其他web服务，可是安装了xray科学富强后，速度测速只有80Mmps，请问：这是nginxSNI这项技术的弊端吗，还是说我猜测可以实现相同的速度，只是多了这一步监听转发对vps性能过高，但我检查了CPU，两种情况都没有超过50%的负荷，所以特别想知道各位使用了分流共用443端口后，速度是否有明显的减弱。

我也发现这个问题了，使用nginx前置通过sni分流到不同的xray节点，发现速度比xray前置速度慢很多，几乎有6/7倍那么多。现在我也猜想是nginx前置的问题。请问楼主解决nginx前置速度下降这个问题了吗？

**Reply by MoRanYue** [reply]
> > 我目前使用的是VLESS+TCP+xtls-rprx-direct，网上找到的一键脚本和按步骤搭建，xray都需要占有443端口， 于是找到了利用SNI分流，stream模块加持下，解决了443端口被xray独占的情况, 但是网速大打折扣，比如若直接让xray监听443 速度快，可以跑满我本地的300Mbps电信带宽，但是不能再部署其他的网站，vps资源被浪费了有点不值得；若用了nginx stream模块，监听其他端口，443端口转发后，虽然也可以搭建其他web服务，可是安装了xray科学富强后，速度测速只有80Mmps，请问：这是nginxSNI这项技术的弊端吗，还是说我猜测可以实现相同的速度，只是多了这一步监听转发对vps性能过高，但我检查了CPU，两种情况都没有超过50%的负荷，所以特别想知道各位使用了分流共用443端口后，速度是否有明显的减弱。
> 
> 我也发现这个问题了，使用nginx前置通过sni分流到不同的xray节点，发现速度比xray前置速度慢很多，几乎有6/7倍那么多。现在我也猜想是nginx前置的问题。请问楼主解决nginx前置速度下降这个问题了吗？

（这里亦使用Nginx反向代理，由于并非于其它主机 部署Nginx（Nginx与Xray-core位于同机器中），为了防止不必要的性能开销，这里使用“Unix域套接字”作为跨进程通信方法呀）

**Reply by taxtigerno1** [reply]
我是使用端口来sni分流，但是你上面不是说套接字与用端口看不出差别吗？使用套接字有用吗？

**Reply by MoRanYue** [reply]
> 我是使用端口来sni分流，但是你上面不是说套接字与用端口看不出差别吗？使用套接字有用吗？

（emm，使用了哪个Nginx版本呢？这里所使用的版本，为“Nginx v1.28.0”呀）

**Comment by Ardentwheel** [comment]
这里有说

> 而开启reuseport后（C 图），有多个 listener 共同 bind/listen 相同的 IP/PORT，也就是说每个进程/线程有一个独立的 listener，相当于每个进程/线程独享一个 listener 的全连接队列，新的连接请求由内核hash分配，不需要多个进程/线程竞争某个公共资源，能充分利用多核，减少竞争的资源消耗，效率自然提高了。
> 
> 但同时也是由于这个分配机制，避免了上下文切换，在服务压力不大，网络情况足够好的情况下，进程/线程更有可能专注于持续读取某个慢连接数据而忽视快连接建立的请求，从而造成快连接方卡顿。

https://www.modb.pro/db/1696009590198587392
